internal/controller/controllers/agentclusterinstall_controller.go

```diff
@@ -0,0 +1,129 @@
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// AgentClusterInstallReconciler reconciles a AgentClusterInstall object
+type AgentClusterInstallReconciler struct {
+	client.Client
+	Log              logrus.FieldLogger
+	CRDEventsHandler CRDEventsHandler
+}
+
+// +kubebuilder:rbac:groups=extensions.hive.openshift.io,resources=agentclusterinstalls,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=extensions.hive.openshift.io,resources=agentclusterinstalls/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=extensions.hive.openshift.io,resources=agentclusterinstalls/finalizers,verbs=update
+
+func (r *AgentClusterInstallReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctx := addRequestIdIfNeeded(origCtx)
+	log := logutil.FromContext(ctx, r.Log).WithFields(
+		logrus.Fields{
+			"agent_cluster_install":           req.Name,
+			"agent_cluster_install_namespace": req.Namespace,
+		})
+
+	defer func() {
+		log.Info("AgentClusterInstall Reconcile ended")
+	}()
+
+	log.Info("AgentClusterInstall Reconcile started")
+
+	// Retrieve AgentClusterInstall
+	clusterInstall := &hiveext.AgentClusterInstall{}
+	if err := r.Get(ctx, req.NamespacedName, clusterInstall); err != nil {
+		log.WithError(err).Errorf("Failed to get resource %s", req.NamespacedName)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Retrieve ClusterDeployment
+	clusterDeploymentKubeKey := types.NamespacedName{
+		Namespace: clusterInstall.Namespace,
+		Name:      clusterInstall.Spec.ClusterDeploymentRef.Name,
+	}
+	clusterDeployment := &hivev1.ClusterDeployment{}
+	if err := r.Get(ctx, clusterDeploymentKubeKey, clusterDeployment); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			log.WithError(err).Error(fmt.Sprintf(
+				"failed to get ClusterDeployment with name '%s' in namespace '%s'",
+				clusterDeploymentKubeKey.Name, clusterDeploymentKubeKey.Namespace))
+			return ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}, nil
+		}
+
+		apiErr := errors.New(fmt.Sprintf(
+			"ClusterDeployment with name '%s' in namespace '%s' not found. Ensure that the CRD is already applied.",
+			clusterDeploymentKubeKey.Name, clusterDeploymentKubeKey.Namespace))
+		err = common.NewApiError(http.StatusBadRequest, apiErr)
+		log.Error(err)
+
+		// Set conditions
+		clusterSpecSynced(clusterInstall, err)
+		setClusterConditionsUnknown(clusterInstall)
+
+		if updateErr := r.Status().Update(ctx, clusterInstall); updateErr != nil {
+			log.WithError(updateErr).Error("failed to update AgentClusterInstall Status")
+			return ctrl.Result{Requeue: true}, nil
+		}
+	}
+
+	return ctrl.Result{}, nil
```

Comment:
If the controller was able to fetch a `clusterDeployment` for that `AgentClusterInstall`, it will print nothing and end the iteration.

It might make it unclear to the person who reads the log, what exactly happened here.
The user will see:
`AgentClusterInstall Reconcile started` 
immediately followed by
`AgentClusterInstall Reconcile ended`

I would suggest adding an info log saying that a clusterDeployment resource was detected for that `AgentClusterInstall` resource, so the rest of the handling will be done by the clusterDeployments controller.